pkg/controller/statefulset/stateful_set_control.go

```diff
@@ -574,35 +576,96 @@ func (ssc *defaultStatefulSetControl) updateStatefulSet(
 		return &status, nil
 	}
 
+	return updateStatefulSetAfterInvariantEstablished(ctx,
+		ssc,
+		set,
+		replicas,
+		updateRevision,
+		status,
+	)
+
+}
+
+func updateStatefulSetAfterInvariantEstablished(
```

Comment:
My thoughts on this function -- this is a lot of logic change. If it goes wrong even when the feature isn't used, how do users disable it? It doesn't check the feature flag.

I would keep the old version of this logic, and decide whether to use the old or new version by whether the flag is on or not.

Reply:
@lavalamp  this function was added/extracted from previous function,  based on earlier comments, to make the new logic more testable. There is a table driven unit test that exercises it

Reply:
Yes, I understand what is going on and I was not complaining about that.

My issue is that it's also a logic change (it always performs a calculation using a maxUnavailable value) and if it is wrong, you can't fix it by turning the flag off.

Reply:
understood, i guess that will be a lot of duplicate code. I am open to it , let me know if that is a blocker ? @soltysh  thoughts ?

Reply:
Not sure if this is a blocker considering it is updating status, the controller may not react to the changes but I agree with @lavalamp that it may cause confusion if the newly added code goes wrong to the end-user or to the entities looking at the status. @krmayankk how much of work would it be to have both the old logic and new logic gated by featuregate flag?

Reply:
If you can convince me that the new logic is really minimal, or that the old tests were exhaustive and still pass, I might be able to hold my nose. I am not sure if it will be easier to convince me or @liggitt but I think you should convince one of us.

Reply:
Resolving since the duplication isn't that bad.